File: TicTacToeGame.cpp
#include <iostream>
#include "TicTacToeGame.h"
#include "Player.h"

Turn TicTacToeGame::nextTurn() {
    Turn turn;
    while (true) {
        turn = m_players[m_currentPlayer]->makeTurn();
        if (m_grid.isValidCell(turn.row, turn.column))
            break;
        else
            std::cout << "Given cell is occupied, please try another one" << std::endl << std::endl;
    }
    turn.index = m_currentTurnIndex++;
    m_grid.put(turn.row, turn.column, (unsigned char) turn.playerId);
    checkGameOver(turn);
    m_currentPlayer = (short) (m_currentPlayer == NUM_PLAYERS - 1 ? 0 : m_currentPlayer + 1);
    return turn;
}

TicTacToeGame::~TicTacToeGame() {
    for (int i = NUM_PLAYERS - 1; i >= 0; --i) {
        delete m_players[i];
    }
}

void TicTacToeGame::initGrid() {
    for (int i = 0; i < NUM_BLOCKED_CELLS; ++i) {
        auto cell = m_randomCellMaker.generate();
        m_grid.put(cell.first, cell.second, 0);
    }
}

void TicTacToeGame::checkGameOver(Turn turn) {
    if (m_grid.makesLine(turn.row, turn.column)) {
        m_winner = turn.playerId;
        return;
    }
    if (turn.index >= Grid::SIZE * Grid::SIZE - NUM_BLOCKED_CELLS - 1)
        m_winner = 0;
}

File: TicTacToeGame.lua
require "RandomCellCoordinateGenerator"
require "PlayersInitializer"
require "Grid"
require "Turn"

TicTacToeGame = {}

NUM_PLAYERS = 3

local NUM_BLOCKED_CELLS = 4

function TicTacToeGame:new(numHumans)
    local o = {}
    setmetatable(o, self)
    self.__index = self
    self._winner = -1
    self._currentPlayer = 1
    self._currentTurnIndex = 0
    self._players = {size = NUM_PLAYERS}
    self._grid = Grid:new()
    self._cellCoordinateGenerator = RandomCellCoordinateGenerator:new(self._grid)
    PlayersInitializer.initialize(self, numHumans)
    self:initGrid()
    return o
end

function TicTacToeGame:initGrid()
    for i = 1, NUM_BLOCKED_CELLS do
        local row, column = self._cellCoordinateGenerator:generate()
        self._grid:put(row, column, 0)
    end
end

function TicTacToeGame:winner()
    return self._winner
end

function TicTacToeGame:checkGameOver(turn)
    if (self._grid:makesLine(turn.row, turn.column)) then
        self._winner = turn.playerId
        return
    end
    if (turn.index >= GRID_SIZE * GRID_SIZE - NUM_BLOCKED_CELLS - 1) then
        self._winner = 0
    end
end

function TicTacToeGame:grid()
    return self._grid
end

function TicTacToeGame:isOver()
    return self._winner ~= -1
end

function TicTacToeGame:nextTurn()
    local turn
    while (true) do
        turn = self._players[self._currentPlayer]:makeTurn()
        if (self._grid:isValidCell(turn.row, turn.column)) then
            break
        else
            print("Given cell is occupied, please try another one\n")
        end
    end

    turn.index = self._currentTurnIndex
    self._currentTurnIndex = self._currentTurnIndex + 1
    self._grid:put(turn.row, turn.column, turn.playerId)
    self:checkGameOver(turn)
    self._currentPlayer = self._currentPlayer == NUM_PLAYERS and 1 or self._currentPlayer + 1
    return turn
end